KCSs/kcs-3.md

```diff
@@ -0,0 +1,310 @@
+
+Protobuf definition:
+
+```proto
+enum direction {
+   ascending = 0;
+   descending = 1;
+}
+
+message spender_value {
+   bytes spender = 1 [(koinos.btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+
+// Arguments
+message get_allowances_args {
+   bytes owner = 1 [(koinos.btype) = ADDRESS];
+   bytes start = 2 [(koinos.btype) = ADDRESS];
+   int32 limit = 3;
+   direction direction = 4;
+}
+
+// Result
+message get_allowances_result {
+   bytes owner = 1 [(koinos.btype) = ADDRESS];
+   repeated spender_value allowances = 2;   
+}
+```
+
+### Write methods
+
+#### mint
+
+Used by the contract owner to initially mint the token to a given address.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message mint_arguments {
+   bytes to = 1 [(btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+// Result
+message mint_result {}
+```
+
+The method should emit a `mint_event` upon success. The event should indicate the recipient of the mint as an impacted account.
+
+```proto
+// Event
+message mint_event {
+   bytes to = 1 [(btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+```
+
+#### transfer
+
+This will transfer tokens to a new owner. The authorization is checked with the native `check_authority` system call. It is also authorized if the contract of `from` is the one that called the token contract.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message transfer_args {
+   bytes from = 1 [(koinos.btype) = ADDRESS];
+   bytes to = 2 [(koinos.btype) = ADDRESS];
+   uint64 value = 3 [jstype = JS_STRING];
+}
+// Result
+message transfer_result {}
+```
+
+The transfer event should emit a `token_transfer` event upon success. The event should indicate the sender and receiver as impacted accounts.
+
+```proto
+// Event
+message transfer_event {
+   bytes from = 1 [(btype) = ADDRESS];
+   bytes to = 2 [(btype) = ADDRESS];
+   uint64 value = 3 [jstype = JS_STRING];
+}
+```
+
+#### burn (optional)
+
+Burns an amount of token from an address. The authorization is checked with the native `check_authority` system call. It is also authorized if the contract of `from` is the one that called the token contract.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message burn_args {
+   bytes from = 1 [(koinos.btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+// Result
+message burn_result {}
+```
+
+The method should emit a `burn_event` upon success. The event should indicate the source as an impacted account.
+
+```proto
+// Event
+message burn_event {
+   bytes from = 1 [(btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+```
+
+#### approve
+
+Grant permissions to other account to manage the tokens owned by the user.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message approve_args {
+   bytes owner = 1 [(koinos.btype) = ADDRESS];
+   bytes spender = 2 [(koinos.btype) = ADDRESS];
+   uint64 value = 3 [jstype = JS_STRING];
+}
+// Result
+message approve_result {}
```

Comment:
add an event

Reply:
We should add these standards/events in the `koinos-proto` repo and align them with standards. I'll look at your event in your token example.

Reply:
```proto
message approve_event {
   bytes owner = 1 [(koinos.btype) = ADDRESS];
   bytes spender = 2 [(koinos.btype) = ADDRESS];
   uint64 value = 3 [jstype = JS_STRING];
}
```

Seems reasonable to me.

Reply:
Updated.